### ISSUE TITLE
Switch AWS SDK to v3

### ISSUE THREAD

**Issue by m-courtin** [serverless/serverless#11920]
### Is there an existing issue for this?

- [X] I have searched existing issues, it hasn't been reported yet

### Use case description

Dear Maintainers, Dear Supporters,

we are currently using the serverless and are now migrating towards node18 -> near support end of node 16 and towards AWS SDK v3, as v2 will be put to maintenance soon. Your latest code is currently using the AWS SDK V2.x 

Do you have plans when to release a version of the serverless using the AWS SDK V3 ?

Thank you very much in advance for your reply
and best regards,

Michael

**Comment by webjunkie** [comment]
Here I found one mention. Seems it's not really on the radar.

https://github.com/serverless/serverless/issues/11753#issuecomment-1436987593

**Comment by theRocket** [comment]
A release was put out to suppress the warning from AWS SDK. https://github.com/serverless/serverless/releases/tag/v3.28.1

**Comment by medikoo** [comment]
As mentioned at https://github.com/serverless/serverless/issues/11753#issuecomment-1436987593 we're open for the PR.

I believe it can be done that way:
 - Upgrade [`lib/aws/request`](https://github.com/serverless/serverless/blob/75ce58b194f19b4fc5a37a7a437733befefccf06/lib/aws/request.js) to support direct injection of v3 Clients alternatively of service names (similar to how it's done here: https://github.com/serverless/test/blob/61899a6dd894539fd7a55a6504e1db1dea35df4e/aws-request.js)
 - Upgrade [`provide.request`](https://github.com/serverless/serverless/blob/75ce58b194f19b4fc5a37a7a437733befefccf06/lib/plugins/aws/provider.js#L1683) to support AWS SDK v3 clients alternatively to service names
 - Upgrade all Framework logic to pass AWS SDK v3 clients to `provider.request`

Having that we will switch Framework to rely on AWS SDK v3 solely, then this work could be followed up with:
 - Remove warning prevention from AWS SDK v2 require, and require it conditionally (only on explicit usage)
 - Introduce deprecation notice when AWS SDK v2 is required - having that we can remove it from dependencies with the major
 
 PR's welcome, let me know if anything is not clear

**Comment by reshab48** [comment]
Hi @medikoo, I am trying to get started with contributing and wanted to work on this issue. Can I jump into the implementation and create a PR?

**Comment by medikoo** [comment]
@reshab48 definitely, PR's is welcome! (still note this can turn challenging and possibly time taking, as we rely on AWS SDK v2 in many places)

**Comment by reshab48** [comment]
@medikoo I have gone through the requirements and I am interested to work on this issue. Can you please assign it to me?

I need some more clarification, so by saying _direct injection of v3 Clients alternatively of service names_ you mean it will require injection of client service instance directly instead of dynamically getting it using lodash right?

And as I understand the last 2 points (followups) should be done as a separate PR once we confirm all clients are migrated to v3?

**Comment by medikoo** [comment]
@reshab48 great thanks for interest 👍 Such contribution will be extremely valuable to the project

> I need some more clarification, so by saying direct injection of v3 Clients alternatively of service names you mean it will require injection of client service instance directly instead of dynamically getting it using lodash right?

Not exactly, wIth AWS SDK v3 to obtain clients by name, we will have to rely dynamic requires (as e.g. `@aws-sdk/client-${clientName}`), and that's a bad pattern (putting aside that it may introduce other maintainance issues).

So instead we expect consumers to _require_ client on their own. We already follow that pattern in tests (which use util similar to `aws/request` but which is hosted in `@serverless/test` project:  https://github.com/serverless/serverless/blob/4d64730130c44cdfbb872f8a16111df409135dc8/test/integration/aws/api-gateway-external.test.js#L23-L25

> And as I understand the last 2 points (followups) should be done as a separate PR once we confirm all clients are migrated to v3?

Yes, e.g. we also need to ensure that core functionality which is placed in outer packages, also doesn't use AWS SDK v2, e.g.:  https://github.com/serverless/dashboard-plugin (and it could be the only one to update).

@reshab48 probably best way to start with it, is if you prepare a PR with `lib/aws/request` , `provider.request` method updated and some single AWS SDK call in core plugins refactored. I'll review that, and once we have agreement it's all good, you may follow with refactor of all other internal AWS SDK calls

**Comment by medikoo** [comment]
@billios I'm no longer with Serverless Inc., so I'm not sure what are the plans towards having that. I believe @Danwakeem is now the right person to ask

**Comment by billios** [comment]
@Danwakeem Is there any estimation for when to release a version of the serverless using the AWS SDK V3?

**Comment by reshab48** [comment]
@medikoo I am almost ready with the updates, but to make this migration work we need to use non-modular v2 like interface to be compatible with the existing architecture. I need to finalize my updates and test, will post my updates here in the upcoming week.

**Comment by medikoo** [comment]
@reshab48 thanks for update, still note I'm no longer with Serverless Inc, and I don't have any rights to push any work here forward. Please consult any further steps with @Mmarzex and @ac360

**Comment by billios** [comment]
Is there any plan to release a version of the serverless using the AWS SDK V3? 
Otherwise, we have to include the AWS SDK V2 ( for Node.js version >= 18) resulting in quite big lambda bundles, just for using serverless.

**Comment by jounii** [comment]
> Is there any plan to release a version of the serverless using the AWS SDK V3? Otherwise, we have to include the AWS SDK V2 ( for Node.js version >= 18) resulting in quite big lambda bundles, just for using serverless.

There should be no reason to include serverless in lambda bundle, unless you need the serverless in your lambda. Serverless should dev dependency. So yes, you need the aws-sdk v2 in dev and CI/CD environments, but not in the runtimes in lambda. There you can use aws-sdk v3 fine.

**Comment by gavinjohnson-crowdcube** [comment]
Bump on this, come July/August anyone who's stacks have the custom resource helper lambdas are not going to be able to deploy anymore as AWS drops support for nodejs16.x

**Comment by JoshuaKemmerer** [comment]
> @ medikoo I am almost ready with the updates, but to make this migration work we need to use non-modular v2 like interface to be compatible with the existing architecture. I need to finalize my updates and test, will post my updates here in the upcoming week.

@reshab48 have you had a chance to make progress on upgrading to use AWS SDK v3?

**Comment by eliasjtg** [comment]
Any update? Workaround?

**Comment by trivikr** [comment]
AWS SDK for JavaScript v2 will enter maintenance mode on September 8, 2024 and reach end-of-support on September 8, 2025. For more information, check blog post at https://a.co/cUPnyil

**Comment by trivikr** [comment]
The README points out that Serverless V4 is in Beta.

Since Serverless V3 was released in Jan'22. The AWS SDK for JavaScript v3 migration should ideally be considered before RC release of Serverless V4, unless there's a way to update it without breaking serverless customers.

cc recent contributors @austencollins @Mmarzex @eahefnawy @czubocha

**Comment by tasnim0tantawi** [comment]
Any updates @reshab48 ?

**Comment by adminy** [comment]
I recommend switching to cdk or terraform. Its not worth it to stick to serverless as its full of bugs and slowly dying framework.

**Comment by sprengerst** [comment]
+1 :)

**Comment by czubocha** [comment]
We're closing this issue since support for AWS SDK for JavaScript v3 is available starting in [v4.19.0](https://github.com/serverless/serverless/releases/tag/v4.19.0). You can enable it by setting the environment variable `SLS_AWS_SDK=3`. See the [docs](https://www.serverless.com/framework/docs/providers/aws/guide/deploying#aws-sdk-for-javascript-v3) for more details.
